### ISSUE TITLE
Disassemble+decompile shows reflective calls that clojure.core/compile doesn't make

### ISSUE THREAD

**Issue by favila** [clojure-goes-fast/clj-java-decompiler#10]
There appear to be differences (at least related to reflection) between what disassemble and decompile show and the class that `clojure.core/compile` actually produces.
I am using clojure 1.11.1 and clj-java-decompiler 0.3.1.

This is a minimal test namespace:

```clojure
(ns x.decompile)

(def M {:a 1 :b 2 :c 3})

(defn inline-all [] (.intValue ^Long (get M :a)))

(defn outline-get []
  (let [l ^Long (get M :a)]
    (.intValue l)))

(defn outline-all []
  (let [^Long l (get M :a)]
    (.intValue l)))
```

I expected Clojure to produce similar, non-reflective code for all three, but `disassemble` and `decompile` say that `inline-all` and `outline-get` both act like they don't see the `^Long` type hint and include a call to `Reflector.invokeNoArgInstanceMember()`.

This is the output from `disassemble`:

```
clojure -Sdeps '{:deps {com.clojure-goes-fast/clj-java-decompiler {:mvn/version "0.3.1"}}}'  --repl
WARNING: Implicit use of clojure.main with options is deprecated, use -M
Clojure 1.11.1
user=> (require 'clj-java-decompiler.core)
(require 'x.decompile)
(in-ns 'x.decompile)
(binding [*compiler-options* {:disable-locals-clearing false}]
  (clj-java-decompiler.core/disassemble
    (defn inline-all [] (.intValue ^Long (get M :a)))))
nil
user=> nil
user=> #object[clojure.lang.Namespace 0x652ab8d9 "x.decompile"]
x.decompile=> 
// Decompiling class: x/decompile$inline_all
class x.decompile$inline_all
    Minor version: 0
    Major version: 52
    Flags: PUBLIC, FINAL, SUPER
    
    public static final clojure.lang.Var const__1;
        Flags: PUBLIC, STATIC, FINAL
    
    public static final clojure.lang.Keyword const__2;
        Flags: PUBLIC, STATIC, FINAL
    
    public void <init>();
        Flags: PUBLIC
        Code:
                  linenumber      1
               0: aload_0        
               1: invokespecial   clojure/lang/AFunction.<init>:()V
               4: return         
    
    public static java.lang.Object invokeStatic();
        Flags: PUBLIC, STATIC
        Code:
                  linenumber      1
               0: getstatic       x/decompile$inline_all.const__1:Lclojure/lang/Var;
               3: invokevirtual   clojure/lang/Var.getRawRoot:()Ljava/lang/Object;
               6: getstatic       x/decompile$inline_all.const__2:Lclojure/lang/Keyword;
                  linenumber      1
               9: invokestatic    clojure/lang/RT.get:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
                  linenumber      1
              12: ldc             "intValue"
              14: iconst_0       
              15: invokestatic    clojure/lang/Reflector.invokeNoArgInstanceMember:(Ljava/lang/Object;Ljava/lang/String;Z)Ljava/lang/Object;
              18: areturn        
    
    public java.lang.Object invoke();
        Flags: PUBLIC
        Code:
                  linenumber      1
               0: invokestatic    x/decompile$inline_all.invokeStatic:()Ljava/lang/Object;
               3: areturn        
    
    static {};
        Flags: PUBLIC, STATIC
        Code:
                  linenumber      1
               0: ldc             "x.decompile"
               2: ldc             "M"
               4: invokestatic    clojure/lang/RT.var:(Ljava/lang/String;Ljava/lang/String;)Lclojure/lang/Var;
               7: checkcast       Lclojure/lang/Var;
              10: putstatic       x/decompile$inline_all.const__1:Lclojure/lang/Var;
              13: aconst_null    
              14: ldc             "a"
              16: invokestatic    clojure/lang/RT.keyword:(Ljava/lang/String;Ljava/lang/String;)Lclojure/lang/Keyword;
              19: checkcast       Lclojure/lang/Keyword;
              22: putstatic       x/decompile$inline_all.const__2:Lclojure/lang/Keyword;
              25: return         

nil
x.decompile=> 
```

This is what the clojure compiler produces, according to `javap` inspecting the AOT-ed class file:

```shell
clojure -Sdeps '{:deps {com.clojure-goes-fast/clj-java-decompiler {:mvn/version "0.3.1"}}}'  -e "(compile 'x.decompile)"

javap -c classes/x/decompile\$inline_all.class 
Compiled from "decompile.clj"
public final class x.decompile$inline_all extends clojure.lang.AFunction {
  public static final clojure.lang.Var const__1;

  public static final clojure.lang.Keyword const__2;

  public x.decompile$inline_all();
    Code:
       0: aload_0
       1: invokespecial #9                  // Method clojure/lang/AFunction."<init>":()V
       4: return

  public static java.lang.Object invokeStatic();
    Code:
       0: getstatic     #15                 // Field const__1:Lclojure/lang/Var;
       3: invokevirtual #20                 // Method clojure/lang/Var.getRawRoot:()Ljava/lang/Object;
       6: getstatic     #24                 // Field const__2:Lclojure/lang/Keyword;
       9: invokestatic  #30                 // Method clojure/lang/RT.get:(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;
      12: checkcast     #32                 // class java/lang/Long
      15: invokevirtual #36                 // Method java/lang/Long.intValue:()I
      18: invokestatic  #42                 // Method java/lang/Integer.valueOf:(I)Ljava/lang/Integer;
      21: areturn

  public java.lang.Object invoke();
    Code:
       0: invokestatic  #45                 // Method invokeStatic:()Ljava/lang/Object;
       3: areturn

  public static {};
    Code:
       0: ldc           #48                 // String x.decompile
       2: ldc           #50                 // String M
       4: invokestatic  #54                 // Method clojure/lang/RT.var:(Ljava/lang/String;Ljava/lang/String;)Lclojure/lang/Var;
       7: checkcast     #17                 // class clojure/lang/Var
      10: putstatic     #15                 // Field const__1:Lclojure/lang/Var;
      13: aconst_null
      14: ldc           #56                 // String a
      16: invokestatic  #60                 // Method clojure/lang/RT.keyword:(Ljava/lang/String;Ljava/lang/String;)Lclojure/lang/Keyword;
      19: checkcast     #62                 // class clojure/lang/Keyword
      22: putstatic     #24                 // Field const__2:Lclojure/lang/Keyword;
      25: return
}
```

In fact `javap` reports identical `invokeStatic` bytecode for all three versions except that the `outline_*` have an extra bit in the middle for locals clearing.

```   
      12: astore_0
      13: aload_0
      14: aconst_null
      15: astore_0
```

None of the class files produced by `compile` have any reflection.

I can't account for this difference, and don't know if it's real (i.e. clojure aot is producing something different from eval) or an artifact of clj-java-decompiler.

**Comment by favila** [comment]
This doesn't happen in clj-java-decompiler 0.3.0! I think that's a clue.

I see in https://github.com/clojure-goes-fast/clj-java-decompiler/commit/48279cdbf44145050c6cb00c13e687d555293b69 you introduced a `clojure.walk/walk` transformation over the input form. I *strongly suspect* that this is dropping metadata from seqs. `walk` has a known bug where it doesn't preserve metadata: https://clojure.atlassian.net/browse/CLJ-2568 I've been bitten by this bug several times before...

**Comment by alexander-yakushev** [comment]
Great catch, thanks a lot! Pretty sure I suffered from this before too.

**Comment by alexander-yakushev** [comment]
Should be fixed in `0.3.2`, waiting for your confirmation 😄.

**Comment by favila** [comment]
Seems to work! Thank you!
